File: day02/day02.ts
// const input: string = await Bun.file("input").text();
import { readFileSync } from "node:fs";
const input: string = readFileSync("input", "utf8");

type Reveal = {
  red: number;
  green: number;
  blue: number;
};

type Game = { id: number; reveals: Reveal[] };

function parse_reveal(reveal_str: string): Reveal {
  let r = { red: 0, green: 0, blue: 0 };
  reveal_str
    .split(", ")
    .map((s) => (r[s.split(" ")[1]] = Number(s.split(" ")[0])));
  return r;
}

function parse_game(game_str: string): Game {
  let [game, reveals] = game_str.split(": ");
  let game_id = Number(game.split(" ")[1]);
  let r = reveals.split("; ").map(parse_reveal);
  return { id: game_id, reveals: r };
}

let games = input
  .split("\n")
  .filter((s) => s.length > 0)
  .map(parse_game);

let part1 = (games: Game[]): number =>
  games
    .filter((g) =>
      g.reveals.every((r) => r.red <= 12 && r.green <= 13 && r.blue <= 14),
    )
    .map((g) => g.id)
    .reduce((a, b) => a + b);

let part2 = (games: Game[]): number =>
  games
    .map((g) =>
      g.reveals.reduce(
        (o, r) => ({
          red: Math.max(o.red, r.red),
          green: Math.max(o.green, r.green),
          blue: Math.max(o.blue, r.blue),
        }),
        { red: 0, green: 0, blue: 0 },
      ),
    )
    .map((r) => r.red * r.green * r.blue)
    .reduce((a, b) => a + b);

console.log(part1(games));
console.log(part2(games));

File: day02/day02.py
with open("input") as fd:
    lines = fd.readlines()

games = {}
for line in lines:
    line.split(":")
    game, reveals = line.split(":")
    game_id = int(game.split()[1])
    r = []
    for reveal in reveals.split(";"):
        d = {}
        for color in reveal.split(","):
            n, col = color.split()
            d[col] = int(n)
        r.append(d)
    games[game_id] = r


def check_game(reveals, thresh):
    for reveal in reveals:
        for color in ("red", "green", "blue"):
            if reveal.get(color, 0) > thresh[color]:
                return False
    return True

thresh = {"red": 12, "green": 13, "blue": 14}
r = 0
for game_id, reveals in games.items():
    if check_game(reveals, thresh):
        r += game_id
print(r)

def min_games(reveals):
    return {
        color: max(r.get(color, 0) for r in reveals)
        for color in ("red", "green", "blue")
    }

r = 0
for _, reveals in games.items():
    x, y, z = min_games(reveals).values()
    r += x * y * z
print(r)
